From 2c02df12fff9958962210565fbe39329647405f6 Mon Sep 17 00:00:00 2001
From: Cheick Traore <cheick.traore@foss.st.com>
Date: Mon, 16 Sep 2024 15:44:09 +0200
Subject: [PATCH 1527/1695] plat-stm32mp2: replace alarm by events for hse
 monitoring

Use events instead of alarm to be triggered when HSE clock
is overclocking.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 5a1920f7f35d2d13adf61dc2b6567c42540255dc plat-stm32mp2: replace alarm by events for hse monitoring]
Change-Id: I589428b996ff2787fec13ed7e631c7146e7036fe
Signed-off-by: Cheick Traore <cheick.traore@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/404183
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c | 59 ++++++++++++++++++++++--------
 1 file changed, 44 insertions(+), 15 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index ee4d10183..6b58a2c3c 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -310,19 +310,22 @@ void __noreturn do_reset(const char *str __maybe_unused)
 }
 
 #ifdef CFG_STM32_HSE_MONITORING
-/* pourcent rate of hse alarm */
-#define HSE_ALARM_PERCENT	110
+/* Percent rate of HSE event */
+#define HSE_THRESHOLD_PERCENT	110
 #define FREQ_MONITOR_COMPAT	"st,freq-monitor"
 
 struct stm32_hse_monitoring_data {
 	struct counter_device *counter;
 	void *config;
+	uint32_t threshold;
 };
 
-static void stm32_hse_over_frequency(uint32_t ticks __unused,
-				     void *user_data __unused)
+static void stm32_hse_over_frequency(void *priv,
+				     enum counter_event_type event __unused)
 {
-	EMSG("HSE over frequency: nb ticks:%"PRIu32, ticks);
+	uint32_t __maybe_unused *ticks = priv;
+
+	EMSG("HSE over frequency detected: nb ticks:%"PRIu32, *ticks);
 }
 DECLARE_KEEP_PAGER(stm32_hse_over_frequency);
 
@@ -332,13 +335,31 @@ static TEE_Result stm32_hse_monitoring_pm(enum pm_op op,
 {
 	struct stm32_hse_monitoring_data *priv =
 	(struct stm32_hse_monitoring_data *)PM_CALLBACK_GET_HANDLE(h);
+	TEE_Result res = TEE_ERROR_GENERIC;
 
 	if (op == PM_OP_RESUME) {
-		counter_start(priv->counter, priv->config);
-		counter_set_alarm(priv->counter);
+		res = counter_set_threshold(priv->counter, priv->threshold);
+		if (res)
+			return res;
+
+		res = counter_enable_event(priv->counter,
+					   COUNTER_EVENT_THRESHOLD,
+					   stm32_hse_over_frequency,
+					   (void *)&priv->threshold);
+		if (res)
+			return res;
+
+		res = counter_start(priv->counter, priv->config);
+		if (res)
+			return res;
 	} else {
-		counter_cancel_alarm(priv->counter);
-		counter_stop(priv->counter);
+		res = counter_stop(priv->counter);
+		if (res)
+			return res;
+
+		res = counter_disable_all_events(priv->counter);
+		if (res)
+			return res;
 	}
 
 	return TEE_SUCCESS;
@@ -387,7 +408,7 @@ static TEE_Result stm32_hse_monitoring(void)
 	 */
 	hsi_cal /= 1024;
 
-	ticks = (hse / 100) * HSE_ALARM_PERCENT;
+	ticks = (hse / 100) * HSE_THRESHOLD_PERCENT;
 	ticks /= hsi_cal;
 
 	DMSG("HSE:%luHz HSI cal:%luHz alarm:%"PRIu32, hse, hsi_cal, ticks);
@@ -395,17 +416,25 @@ static TEE_Result stm32_hse_monitoring(void)
 	counter = fdt_counter_get(fdt, node, &config);
 	assert(counter && config);
 
-	counter->alarm.callback = stm32_hse_over_frequency;
-	counter->alarm.ticks = ticks;
-
 	priv->counter = counter;
 	priv->config = config;
+	priv->threshold = ticks;
 
 	register_pm_core_service_cb(stm32_hse_monitoring_pm, priv,
 				    "stm32-hse-monitoring");
+	res = counter_set_threshold(counter, priv->threshold);
+	if  (res)
+		return res;
+
+	res = counter_enable_event(counter, COUNTER_EVENT_THRESHOLD,
+				   stm32_hse_over_frequency,
+				   (void *)&priv->threshold);
+	if  (res)
+		return res;
 
-	counter_start(counter, config);
-	counter_set_alarm(counter);
+	res = counter_start(counter, config);
+	if  (res)
+		return res;
 
 	return TEE_SUCCESS;
 }
-- 
2.47.2

